modules/image-loading-optimization/module.php

```diff
@@ -0,0 +1,299 @@
+		if ( ! isset( $attr['width'], $attr['height'] ) ) {
+			return $loading_attrs;
+		}
+
+		/*
+		 * The key function logic starts here.
+		 */
+		$maybe_in_viewport    = null;
+		$increase_count       = false;
+		$maybe_increase_count = false;
+
+		// Logic to handle a `loading` attribute that is already provided.
+		if ( isset( $attr['loading'] ) ) {
+			/*
+			 * Interpret "lazy" as not in viewport. Any other value can be
+			 * interpreted as in viewport (realistically only "eager" or `false`
+			 * to force-omit the attribute are other potential values).
+			 */
+			if ( 'lazy' === $attr['loading'] ) {
+				$maybe_in_viewport = false;
+			} else {
+				$maybe_in_viewport = true;
+			}
+		}
+
+		// Logic to handle a `fetchpriority` attribute that is already provided.
+		if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
+			/*
+			 * If the image was already determined to not be in the viewport (e.g.
+			 * from an already provided `loading` attribute), trigger a warning.
+			 * Otherwise, the value can be interpreted as in viewport, since only
+			 * the most important in-viewport image should have `fetchpriority` set
+			 * to "high".
+			 */
+			if ( false === $maybe_in_viewport ) {
+				_doing_it_wrong(
+					__FUNCTION__,
+					__( 'An image should not be lazy-loaded and marked as high priority at the same time.', 'elementor' )
+				);
+				/*
+				 * Set `fetchpriority` here for backward-compatibility as we should
+				 * not override what a developer decided, even though it seems
+				 * incorrect.
+				 */
+				$loading_attrs['fetchpriority'] = 'high';
+			} else {
+				$maybe_in_viewport = true;
+			}
+		}
+
+		if ( null === $maybe_in_viewport ) {
+			if ( ! is_admin() ) {
+				$content_media_count = $this->increase_content_media_count( 0 );
+				$increase_count      = true;
+				if ( $content_media_count < $this->omit_loading_attr_threshold() ) {
+					$maybe_in_viewport = true;
+				} else {
+					$maybe_in_viewport = false;
+				}
+			}
+		}
+
+		if ( $maybe_in_viewport ) {
+			$loading_attrs = $this->maybe_add_fetchpriority_high_attr( $loading_attrs, $tag_name, $attr );
+		} else {
+			$loading_attrs['loading'] = 'lazy';
+		}
+
+		if ( $increase_count ) {
+			$this->increase_content_media_count();
+		} elseif ( $maybe_increase_count ) {
+			if ( $this->min_priority_img_pixels <= $attr['width'] * $attr['height'] ) {
+				$this->increase_content_media_count();
+			}
+		}
+
+		return $loading_attrs;
+	}
+
+	private function increase_content_media_count( $amount = 1 ) {
+		static $content_media_count = 0;
+
+		$content_media_count += $amount;
+
+		return $content_media_count;
+	}
+
+	private function omit_loading_attr_threshold( $force = false ) {
+		static $omit_threshold;
+		if ( ! isset( $omit_threshold ) || $force ) {
+			$omit_threshold = 3;
+		}
+
+		return $omit_threshold;
+	}
+
+	private function maybe_add_fetchpriority_high_attr( $loading_attrs, $tag_name, $attr ) {
```

Comment:
Consider removing the unused `$tag_name` param, per my prior comment

```suggestion
	private function maybe_add_fetchpriority_high_attr( $loading_attrs, $attr ) {
```

Reply:
Fixed.